### PR TITLE
Refactor/todos

### PR DESCRIPTION
**Pull request by tombiddulph**
Actioned all the `TODO Tom` comments

- Switched from Cosmonaut to Entity Framework core
- Configured the EF db as the InMemoryVersion, a production configuration will need to be added in due course
- Moved `ValidatePaymentAmount()` from the smart request repository to the domain model
- Created `IUser`, `PlatformUser` & `User` models
- Changed YapilyUserBuilder method `CreateIfNotExists` to `CreateYapilyUser`
- Rewrote tests for new orm.

### CODE REVIEW

File: src/Domain/User/Model/IParticipant.cs [hunk 1]
```diff
@@ -1,7 +1,11 @@
-﻿namespace Ordo.OpenBankingService.Domain.User.Model
+﻿using System;
```
Comment (hwarrior) [inline src/Domain/User/Model/IParticipant.cs]: This interface is not used anywhere, could you delete it please?

File: src/Api/Model/InitiatePaymentRequest.cs [hunk 2]
```diff
@@ -5,20 +5,31 @@
 
 namespace Ordo.OpenBankingService.Api.Model
 {
-    public class InitiatePaymentRequest : IValidatableObject
+
+    interface IInitiatePaymentRequest
```
Comment (hwarrior) [inline src/Api/Model/InitiatePaymentRequest.cs]: Why do we need this interface? It is not referenced from anywhere in the code, so we are not using it atm.

File: src/Api/Api.csproj [hunk 3]
```diff
@@ -24,7 +24,9 @@
     <PackageReference Include="idunno.Authentication.Basic" Version="2.1.1" />
     <PackageReference Include="Microsoft.AspNetCore.App" Version="2.2.0" />
     <PackageReference Include="Microsoft.AspNetCore.Mvc.Formatters.Json" Version="2.2.0" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.InMemory" Version="2.2.4" />
```
Comment (hwarrior) [inline src/Api/Api.csproj]: This package is not used from the Api project.

File: src/Api/Api.csproj [hunk 4]
```diff
@@ -24,7 +24,9 @@
...
     <PackageReference Include="Microsoft.VisualStudio.Azure.Containers.Tools.Targets" Version="1.5.4" />
+    <PackageReference Include="Moq.EntityFrameworkCore" Version="2.2.1" />
```
Comment (hwarrior) [inline src/Api/Api.csproj]: This package is not used in the Api project.

File: src/Domain/User/Model/User.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    public class PlatformUser : IUser
+    {
+        public string UserId { get; set; }
+    }
+
+    public class User : IUser, IYapilyUser
+    {
+        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
```
Comment (hwarrior) [inline src/Domain/User/Model/User.cs]: Two comments on this one:
1. the domain shouldn't contain hints or references to the underlying physical implementation of the data store.
2. I'm not sure on this yet but I think the Id shouldn't be database generated, rather the the user id from the platform should serve as primary key.

Lets chat about this later:)

File: src/Domain/User/Model/IParticipant.cs (hunk 1 above)
Comment (tombiddulph) [reply src/Domain/User/Model/IParticipant.cs]: Removed

File: src/Api/Model/InitiatePaymentRequest.cs (hunk 2 above)
Comment (tombiddulph) [reply src/Api/Model/InitiatePaymentRequest.cs]: Removed

File: src/Api/Api.csproj (hunk 3 above)
Comment (tombiddulph) [reply src/Api/Api.csproj]: Removed

File: src/Api/Api.csproj (hunk 4 above)
Comment (tombiddulph) [reply src/Api/Api.csproj]: Removed

File: src/Domain/User/Model/User.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        string UserId { get; set; }
+
+    }
+
+    public class PlatformUser : IUser
+    {
+        public string UserId { get; set; }
+    }
+
+    public class User : IUser, IYapilyUser
```
Comment (hwarrior) [inline src/Domain/User/Model/User.cs]: Okay, I think I get it now:
1. This class should be a specific YapilyUser class and contain only yapily user specific information.
2. the UserId(which I believe refers to the platform user) should not be part of this class
3. Class should also be an entity as it have a lifecycle and identity.
Then this class will be used to during creation of new yapily users (plus persisting the mapping)

File: src/Infrastructure/Repositories/YapilyUserRepository.cs [hunk 7]
```diff
@@ -1,85 +1,81 @@
...
 
-            return result;
-        }
+        // TODO move the logic of checking for existence out from infrastructure into the domain(and probably have just one create method)
+        // TODO handle error cases when new mapping is not successful
+        // TODO handle case when Yapily user already exist but there is no mapping on our end
 
 
-        public async Task<IYapilyUser> GetById(IParticipant participant)
+        public async Task<IUser> GetById<T>(T idParam)
```
Comment (hwarrior) [inline src/Infrastructure/Repositories/YapilyUserRepository.cs]: This is not used by production code atm. Could you delete it please?

File: src/Infrastructure/ServiceConnectors/UserStoreConnector.cs [hunk 8]
```diff
@@ -0,0 +1,7 @@
+namespace Ordo.OpenBankingService.Infrastructure.ServiceConnectors
```
Comment (hwarrior) [inline src/Infrastructure/ServiceConnectors/UserStoreConnector.cs]: Unused class?

File: src/InfrastructureTest/YapilyUserRepositoryTests.cs [hunk 9]
```diff
@@ -1,205 +1,163 @@
...
+                UserId = id.ToString()
             };
 
 
-            var expectedUser = new YapilyUser
-            {
-                PlatformId = id.ToString(),
-                Participant = participant
-            };
+            //var expectedUser = new User
```
Comment (hwarrior) [inline src/InfrastructureTest/YapilyUserRepositoryTests.cs]: Remove comment please.

File: src/Domain/User/Model/User.cs (hunk 6 above)
Comment (tombiddulph) [reply src/Domain/User/Model/User.cs]: I've left this for the time being as I think you're working on the same thing

File: src/Domain/User/Model/IYapilyUser.cs [hunk 10]
```diff
@@ -5,8 +5,7 @@ namespace Ordo.OpenBankingService.Domain.User.Model
 {
     public interface IYapilyUser : IAggregateRoot
     {
-        IParticipant Participant { get; set; }
-        // TODO are we using this for anything?
+  
         Guid YapilyApplicationId { get; set; }
```
[the remark this answers is not in the document]
Comment (tombiddulph) [reply src/Domain/User/Model/IYapilyUser.cs]: Removed

File: src/Domain/User/IYapilyUserRepository.cs [hunk 11]
```diff
@@ -8,9 +7,12 @@ namespace Ordo.OpenBankingService.Domain.User
     public interface IYapilyUserRepository : IRepository<YapilyUser>
     {
         // TODO not used by production code atm, remove?
-        Task<IYapilyUser> GetById(Guid id);
-        // TODO rename this method once Participant has been reworked
-        Task<IYapilyUser> GetById(IParticipant participant);
-        Task<IYapilyUser> CreateIfNotExists(IParticipant participant);
+        Task<IUser> GetById<T>(T id);
+
+        Task<IUser> CreateYapilyUser(IUser participant);
```
[the remark this answers is not in the document]
Comment (tombiddulph) [reply src/Domain/User/IYapilyUserRepository.cs]: Changed

File: src/Infrastructure/ServiceConnectors/YapilyServiceConnector.cs [hunk 12]
```diff
@@ -85,11 +85,14 @@ public async Task<IEnumerable<Institution>> GetAll()
         }
 
 
-        public async Task<YapilyCreateUserResponse> CreateUser(IParticipant participant)
+        public async Task<YapilyCreateUserResponse> CreateUser(IUser participant)
```
[the remark this answers is not in the document]
Comment (tombiddulph) [reply src/Infrastructure/ServiceConnectors/YapilyServiceConnector.cs]: Changed

File: src/Infrastructure/Repositories/YapilyUserRepository.cs (hunk 7 above)
Comment (tombiddulph) [reply src/Infrastructure/Repositories/YapilyUserRepository.cs]: Removed

File: src/Infrastructure/ServiceConnectors/UserStoreConnector.cs (hunk 8 above)
Comment (tombiddulph) [reply src/Infrastructure/ServiceConnectors/UserStoreConnector.cs]: Removed

File: src/InfrastructureTest/YapilyUserRepositoryTests.cs (hunk 9 above)
Comment (tombiddulph) [reply src/InfrastructureTest/YapilyUserRepositoryTests.cs]: Removed

File: src/Infrastructure/ServiceConnectors/YapilyServiceConnector.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (hwarrior) [reply src/Infrastructure/ServiceConnectors/YapilyServiceConnector.cs]: Could you also change the name of the parameter please? In addition to the type:)

**Pull request merged**